File: python/0238.product-of-array-except-self/solution.py
# Created by Jones at 2023/11/28 16:42
# leetgo: dev
# https://leetcode.cn/problems/product-of-array-except-self/

from typing import *
from leetgo_py import *

from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from copy import deepcopy
from functools import cache, cmp_to_key, lru_cache, reduce
from heapq import heapify, heappop, heappush, heappushpop, heapreplace
from itertools import accumulate, chain, count, pairwise, zip_longest
from math import ceil, comb, floor, gcd, inf, isqrt, log2, perm, sqrt
from operator import xor
from string import ascii_lowercase
from typing import List, Optional

# @lc code=begin
# from sortedcontainers import SortedList


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        suf = [0] * n
        suf[-1] = nums[-1]
        for i in reversed(range(n - 1)):
            suf[i] = nums[i] * suf[i + 1]
        # for res[n-1]
        suf.append(1)

        res = [0] * n
        left = 1
        for i, x in enumerate(nums):
            res[i] = left * suf[i + 1]
            left *= x
        return res


# @lc code=end

if __name__ == "__main__":
    nums: List[int] = deserialize("List[int]", read_line())
    ans = Solution().productExceptSelf(nums)

    print("\noutput:", serialize(ans))

File: rust/src/0238.product-of-array-except-self/solution.rs
// Created by Jones at 2024/03/15 12:51
// leetgo: 1.4.2
// https://leetcode.com/problems/product-of-array-except-self/

use anyhow::Result;
use leetgo_rs::*;

struct Solution;

// @lc code=begin
use std::cmp::Reverse;
use std::collections::*;
use std::mem::swap;

impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let mut res = vec![1; n];
        (1..n).for_each(|i| res[i] = res[i - 1] * (nums[i - 1]));

        let mut right = nums[n - 1];
        for i in (0..n - 1).rev() {
            res[i] *= right;
            right *= nums[i];
        }
        res
    }
}

// @lc code=end

fn main() -> Result<()> {
    let nums: Vec<i32> = deserialize(&read_line()?)?;
    let ans: Vec<i32> = Solution::product_except_self(nums).into();

    println!("\noutput: {}", serialize(ans)?);
    Ok(())
}
